rust/azure_iot_operations_services/src/leased_lock/lock.rs

```diff
@@ -0,0 +1,177 @@
+    /// - There must be one instance of `lock::Client` per lock.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the either `lock_name` or `lock_holder_name` is empty.
+    pub fn new(
+        state_store: Arc<state_store::Client<C>>,
+        lock_name: Vec<u8>,
+        lock_holder_name: Vec<u8>,
+    ) -> Result<Self, Error> {
+        if lock_name.is_empty() {
+            return Err(Error(ErrorKind::InvalidArgument(
+                "lock_name is empty".to_string(),
+            )));
+        }
+
+        if lock_holder_name.is_empty() {
+            return Err(Error(ErrorKind::InvalidArgument(
+                "lock_holder_name is empty".to_string(),
+            )));
+        }
+
+        let lease_client = lease::Client::new(state_store, lock_name, lock_holder_name)?;
+
+        Ok(Self { lease_client })
+    }
+
+    /// Waits until a lock is available (if not already) and attempts to acquire it.
+    ///
+    /// If a non-zero `Duration` is provided as `renewal_period`, the lock is automatically renewed
+    /// after every consecutive elapse of `renewal_period` until the lock is released or a re-acquire failure occurs.
+    /// If automatic lock renewal is used, `current_lock_fencing_token()` must be used to access the most up-to-date
+    /// fencing token (see function documentation).
+    ///
+    /// Notes:
+    /// `request_timeout` is rounded up to the nearest second.
+    ///
+    /// If lock auto-renewal is used, an auto-renewal task is spawned.
+    /// To terminate this task and stop the lock auto-renewal, `lock::Client::unlock()` must be called.
+    /// Simply dropping the `lock::Client` instance will not terminate the auto-renewal task.
+    /// This logic is intended for a scenario where the `lock::Client` is cloned and a lease is acquired with auto-renewal by the original instance.
+    /// If the original instance is dropped, its clone remains in control of the lock (through the auto-renewal task that remains active).
+    /// Special attention must be used to avoid a memory leak if `lock::Client::unlock()` is never called in this scenario.
+    ///
+    ///
+    /// Returns Ok with a fencing token (`HybridLogicalClock`) if completed successfully, or an `Error` if any failure occurs.
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the `request_timeout` is zero or > `u32::max`
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for the request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    pub async fn lock(
+        &mut self,
```

Comment:
This doesn't need to be mut anymore either as far as I can tell.

Should also then change in sample

Reply:
Fixed.